src/repr/src/scalar.rs

```diff
@@ -1989,6 +1989,42 @@ impl<'a, E, B: DatumType<'a, E>> DatumType<'a, E> for Result<B, E> {
     }
 }
 
+/// A wrapper type that excludes `NULL` values, even if `B` allows them.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ExcludeNull<B>(B);
```

Comment:
Maybe mention in the doc comment that you can use this to make the system auto-determine `propagates_nulls`.